improver/spotdata/height_adjustment.py

```diff
@@ -0,0 +1,207 @@
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Apply's height adjustment for height above ground level spot forecasts."""
+
+from itertools import product
+
+import iris
+import numpy as np
+from iris.cube import Cube
+from iris.exceptions import CoordinateNotFoundError
+from scipy.interpolate import LinearNDInterpolator
+
+from improver import BasePlugin
+from improver.metadata.probabilistic import find_threshold_coordinate, is_probability
+from improver.spotdata.apply_lapse_rate import extract_vertical_displacements
+from improver.utilities.cube_manipulation import enforce_coordinate_ordering
+
+
+class SpotHeightAdjustment(BasePlugin):
+    """
+    Class to adjust spot extracted "height above ground level" forecasts to account
+    for differences between site height and orography grid square height.
+
+    For realization or percentile data the vertical displacement is added on to
+    each realizations or percentile.
+
+    For probability data the data is interpolated between thresholds for each site and
+    the equivalent set of thresholds relative to the site altitude extracted. Any
+    new threshold that is above or below the original set of thresholds uses the highest
+    or lowest threshold's probability from the original cube respectively for that spot.
+    """
+
+    def __init__(self, neighbour_selection_method: str = "nearest",) -> None:
+        """
+        Args:
+            neighbour_selection_method:
+                The neighbour cube may contain one or several sets of grid
+                coordinates that match a spot site. These are determined by
+                the neighbour finding method employed. This keyword is used to
+                extract the desired set of coordinates from the neighbour cube.
+        """
+        self.neighbour_selection_method = neighbour_selection_method
+        self.threshold_coord = None
+        self.units = None
+
+    def adjust_prob_cube(self, spot_cube: Cube, vertical_displacement: Cube) -> Cube:
+        """
+        Adjust probability spot forecasts based on the vertical displacement of sites
+        in relation to orography.
+
+        Args:
+            spot_cube:
+                A cube of spot probability forecasts.
+            vertical_displacement:
+                A cube containing information about the difference between spot
+                data site height and the orography grid square height.
+
+        Returns:
+            A cube with the same metadata and shape as spot_cube but with probabilities
+            adjusted to be relative to the site altitude rather than grid square altitude.
+        """
+        coord_list = [c.name() for c in spot_cube.dim_coords]
+        enforce_coordinate_ordering(
+            spot_cube, [self.threshold_coord.name, "spot_index"]
```

Comment:
Personally I prefer keeping this in the init as it's a bit tidier than having an extra call in the function. This section will only ever ben called if the cube is a probability cube and so it must have a threshold coordinate. None will never be passed into this part. 

As an aside this did also highlight to me that this wasn't actually enforcing the coordinate order the way I wanted it to so I made a slight tweak and added in some extra tests to check the coordinate ordering is working.

Reply:
To avoid `mypy` suggesting the `self.threshold_coord.name` might be an error because it is set as None within the init, you could instead just define `self.threshold_coord = DimCoord`, after importing DimCoord i.e. `from iris.coords import DimCoord`, within the init. This would allow you to keep the current code structure but avoid the `mypy` error.

Reply:
I've followed your suggestion and used the DimCoord. Hopefully that should fix the mypy error.